// @aave/contract-helpers/dist/cjs/v3-faucet-contract/typechain/IERC20FaucetOwnable__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.IERC20FaucetOwnable__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [],
        name: 'isPermissioned',
        outputs: [
            {
                internalType: 'bool',
                name: '',
                type: 'bool',
            },
        ],
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [
            {
                internalType: 'address',
                name: 'token',
                type: 'address',
            },
            {
                internalType: 'address',
                name: 'to',
                type: 'address',
            },
            {
                internalType: 'uint256',
                name: 'amount',
                type: 'uint256',
            },
        ],
        name: 'mint',
        outputs: [
            {
                internalType: 'uint256',
                name: '',
                type: 'uint256',
            },
        ],
        stateMutability: 'nonpayable',
        type: 'function',
    },
    {
        inputs: [
            {
                internalType: 'bool',
                name: 'value',
                type: 'bool',
            },
        ],
        name: 'setPermissioned',
        outputs: [],
        stateMutability: 'nonpayable',
        type: 'function',
    },
];
class IERC20FaucetOwnable__factory {
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.IERC20FaucetOwnable__factory = IERC20FaucetOwnable__factory;
IERC20FaucetOwnable__factory.abi = _abi;
